src/Common/SettingsChanges.h

```diff
@@ -13,11 +14,16 @@ struct SettingChange
     String name;
     Field value;
 
+    /// A setting value which cannot be put in Field.
+    ASTPtr value_ast;
```

Comment:
The idea is nice however `SettingChange` is used in a lot of places and most of those cases don't check `value_ast`. Probably it's better to make `SettingChange` a class and make functions to return either `Field` or `ASTPtr` and check there that the `SettingChange` really contain what is required.